From 00e893e1eefd6616823619b3bb5bf4e7124f8875 Mon Sep 17 00:00:00 2001
From: "Eugene San (eugenesan)" <eugenesan@gmail.com>
Date: Sun, 23 Dec 2012 18:11:29 +0200
Subject: [PATCH] optional encryption password sync

Option to use 2 different tokens for encryption and locking.
Allows convenient usage and strong encryption at the same time.

Inspired by reading next:
https://santoku-linux.com/howto/mobile-forensics/how-to-brute-force-android-encryption

Related discussion:
http://code.google.com/p/android/issues/detail?id=29468

Change-Id: I87f12bb1644f3abcbbd6b23813bbbf2cd58855df

Conflicts:
	core/java/android/provider/Settings.java
	core/java/com/android/internal/widget/LockPatternUtils.java
---
 core/java/android/provider/Settings.java           |  8 +++++++
 .../android/internal/widget/LockPatternUtils.java  | 26 ++++++++++++++++++++--
 .../internal/widget/LockSettingsService.java       |  3 ++-
 3 files changed, 34 insertions(+), 3 deletions(-)

diff --git a/core/java/android/provider/Settings.java b/core/java/android/provider/Settings.java
index 1be054a..80d331e 100644
--- a/core/java/android/provider/Settings.java
+++ b/core/java/android/provider/Settings.java
@@ -3234,6 +3234,7 @@ public final class Settings {
             MOVED_TO_LOCK_SETTINGS.add(Secure.LOCK_PATTERN_ENABLED);
             MOVED_TO_LOCK_SETTINGS.add(Secure.LOCK_PATTERN_VISIBLE);
             MOVED_TO_LOCK_SETTINGS.add(Secure.LOCK_PATTERN_TACTILE_FEEDBACK_ENABLED);
+            MOVED_TO_LOCK_SETTINGS.add(Secure.LOCK_SYNC_ENCRYPTION_PASSWORD);
 
             MOVED_TO_GLOBAL = new HashSet<String>();
             MOVED_TO_GLOBAL.add(Settings.Global.ADB_ENABLED);
@@ -3913,6 +3914,13 @@ public final class Settings {
             "lock_screen_owner_info_enabled";
 
         /**
+         * Whether to sync encryption password with lock screen token
+         * @hide
+         */
+        public static final String LOCK_SYNC_ENCRYPTION_PASSWORD =
+            "lock_sync_encryption_password";
+
+        /**
          * The Logging ID (a unique 64-bit value) as a hex string.
          * Used as a pseudonymous identifier for logging.
          * @deprecated This identifier is poorly initialized and has
diff --git a/core/java/com/android/internal/widget/LockPatternUtils.java b/core/java/com/android/internal/widget/LockPatternUtils.java
index 75fef24..5c8507c 100644
--- a/core/java/com/android/internal/widget/LockPatternUtils.java
+++ b/core/java/com/android/internal/widget/LockPatternUtils.java
@@ -629,8 +629,13 @@ public class LockPatternUtils {
             KeyStore keyStore = KeyStore.getInstance();
             if (password != null) {
                 if (userHandle == UserHandle.USER_OWNER) {
-                    // Update the encryption password.
-                    updateEncryptionPassword(password);
+                    // Sync encryption password if enabled
+                    if (getSyncEncryptionPassword()) {
+                        Log.d(TAG, "Syncing encryption password");
+                        updateEncryptionPassword(password);
+                    } else {
+                        Log.d(TAG, "Skipping encryption password sync");
+                    }
 
                     // Update the keystore password
                     keyStore.password(password);
@@ -1346,4 +1351,21 @@ public class LockPatternUtils {
         return false;
     }
 
+    /**
+     * @hide
+     * Set the lock-sync-encryption-password option (whether to sync encryption
+     * password with lock screen token).
+     */
+    public void setSyncEncryptionPassword(boolean enabled) {
+        setBoolean(Settings.Secure.LOCK_SYNC_ENCRYPTION_PASSWORD, enabled);
+    }
+
+    /**
+     * @hide
+     * Get the lock-sync-encryption-password option (whether to sync encryption
+     * password with lock screen token).
+     */
+    public boolean getSyncEncryptionPassword() {
+        return getBoolean(Settings.Secure.LOCK_SYNC_ENCRYPTION_PASSWORD, false);
+    }
 }
diff --git a/core/java/com/android/internal/widget/LockSettingsService.java b/core/java/com/android/internal/widget/LockSettingsService.java
index 4ecbd16..7643427 100644
--- a/core/java/com/android/internal/widget/LockSettingsService.java
+++ b/core/java/com/android/internal/widget/LockSettingsService.java
@@ -400,6 +400,7 @@ public class LockSettingsService extends ILockSettings.Stub {
         Secure.LOCK_PATTERN_ENABLED,
         Secure.LOCK_BIOMETRIC_WEAK_FLAGS,
         Secure.LOCK_PATTERN_VISIBLE,
-        Secure.LOCK_PATTERN_TACTILE_FEEDBACK_ENABLED
+        Secure.LOCK_PATTERN_TACTILE_FEEDBACK_ENABLED,
+        Secure.LOCK_SYNC_ENCRYPTION_PASSWORD
         };
 }
-- 
1.8.0.3

